### ISSUE TITLE
Feature: Launch Feedback Hub

### ISSUE THREAD

**Issue by mrlacey** [Microsoft/WindowsTemplateStudio#196]
Add the ability to launch the feedback hub to gather user feedback.

Possible functionality:
* On Blank and Pivot&Tabs projects, add to the settings page
* On NavPanel projects add as a link in the SecondaryItems (bottom) menu

Making this a feature, not just an addition to the Settings page as it introduces a new dependency (Microsoft Store Services SDK)

For reference: https://docs.microsoft.com/en-us/windows/uwp/monetize/launch-feedback-hub-from-your-app

**Comment by mrlacey** [comment]
I have a PoC of this and have shown it to Clint & Michael. Will add as a branch next week when I get back from build.

**Comment by crutkas** [comment]
https://docs.microsoft.com/en-us/windows/uwp/monetize/microsoft-store-services-sdk#install-via-nuget for nuget install.

I think we should discuss event logging for VSMC and Store SDK.

Love the ads idea

**Comment by crutkas** [comment]
Should this take a dependency on the settings page?  I feel like yes.

**Comment by mrlacey** [comment]
@crutkas why a dependency on the settings page? Are you thinking it should be something added there? (rather than in the TitleBar?)

**Comment by crutkas** [comment]
My thoughts are what app(s) do that?

**Comment by mrlacey** [comment]
> My thoughts are what app(s) do that?

@crutkas which "that" are you referring to?

**Comment by crutkas** [comment]
doing the feedback hub launch in the titlebar

**Comment by mrlacey** [comment]
There are lots of places this could go:

- Add to the title bar (like Visual Studio)
- Add to the Settings page (like Groove)
- Add to primary navigation menu items. (like the News app)
- Add to the secondary navigation items (above settings)
- Another common place for such a link is an "About Page" (like Calculator) but we don't have one of those.
- As secondary item (hidden behind ellipses) in a command bar at the top of the page (Like Edge & Movies&TV)

Does it have to be in the same place for every project type? I don't think it does.

My proposal for this is:
- If a SplitView project add to the secondary menu (above Settings - I think this highlights how it's different from other menu options which are about navigation within the app)
- If a Pivot&Tabs add to the Command bar at the top of the page. This will, of course, also require adding the ability to add such a command bar (as a hidden feature.)
- Do not include in a blank project. With a blank project, the responsibility for the UI is entirely down to the developer and so they can choose to add a feedback hub link wherever they choose.

Does anyone else have an opinion or preference?

**Comment by mrlacey** [comment]
Did some exploration on this today.
I still like adding the feedback link in the lower list on the nav panel but have hit an issue on the Pivot&Tabs project type.
There isn't a simple way to add it to the top of the Pivot page like I wanted and have it work in a predictable way on narrower windows and/or when there are a lot of tabs.

It looks ok when wide
![image](https://user-images.githubusercontent.com/189547/29241218-5557c292-7f6d-11e7-96a3-227db8904bd9.png)

but crashes into the tab labels when narrow
![image](https://user-images.githubusercontent.com/189547/29241225-7432b730-7f6d-11e7-8521-064ded91eefb.png)

There's no simple way to fix this that's generally applicable

The Store app has something similar
![image](https://user-images.githubusercontent.com/189547/29241234-91af418e-7f6d-11e7-85a9-b7999ebaebba.png)

but this avoids the issue by changing the layout dramatically when narrow
![image](https://user-images.githubusercontent.com/189547/29241236-a027d88e-7f6d-11e7-8342-23dab0e8ad5d.png)

Adding something like that would be overkill for this feature.
I could make it so that the command bar and pivot don't overlap but this would create problems for people down the line who want to put other things in the command bar too as there could end up being competition for space between the tab headers and the commands.

3 options:
1. Just add this feature to the NavPanel project type. Seems like not taking full advantage of the benefit of this feature. I proposed it as gathering user feedback is something every app should do.
2. Add this to the settings page. It creates a dependency on having the settings page though. Is this an issue?
3. Something else. I don't know what and am open to suggestions?

A compromise of 1 & 2 would be to put it on the menu for NavPanel apps but directly in the settings page for blank and tab/pivot apps. Thi is currently my preferred option but aware there's an argument that this could be confusing to some. Although, how many people will generate both project types. Plus if looking for this functionality in the settings/about page then this is no less discoverable.

Feedback, thoughts, and suggestions, please.

**Comment by mrlacey** [comment]
This needs more thought. Bumping to next release

**Comment by mrlacey** [comment]
Removing myself as assignee as don't want to block others from moving this forward.

**Comment by mrlacey** [comment]
It is also for questions like #1133 that I wanted to do this.

**Comment by mrlacey** [comment]
As alluded to in #1133, adding this to the NavPanel menu will mean making the menu take items that don't trigger navigation but execute arbitrary code instead. This is a simple addition to the `ShellNavigationItem` and a modification to ShellViewModel (and equiv.) but makes me think renaming "ShellNavigationItem" to "ShellMenuItem" may be appropriate as it will then be able to do more than just support navigation. Plus, it highlights that the menu can be for more than just navigation too.
Need to consider the impact on the new NavigationView control coming with FCU as well.

**Comment by mrlacey** [comment]
My current thinking is to do this as an addition to the Settings page
- This makes the template simpler and can be the same visually for all project types & frameworks. (This also means it'll be done sooner :wink:)
- #1137 can still be done separately and gets unblocked.
- It avoids making any changes to the Shell just before we introduce NavigationView for FCU. Built-in support for menu items that run code and don't trigger navigation can be added once the integration of NavigationView is stable.
- It'll also introduce chained dependencies (this depends on Settings Page, which depends on Storage) so it'll be good to have something that confirms this is possible. It'll also be the first feature to depend on a page (rather than the other way round) so will be a good test of that.

**Comment by mrlacey** [comment]
As per #827 this should go in the "User Interaction" group

**Comment by mrlacey** [comment]
**Merging with files of possibly changed names.**
Despite my thoughts that I'd found a simple solution to this, I hadn't realized we have no way to merge with a file that may have been renamed.
In this instance merging with SettingsPage.xaml

My plan to address this is:
- Name the new merge file $SEARCH$_postaction.xaml
- Add a new tag to the config like `"wts.fileNameSearch": "x:Uid=\"SettingsPage_Personalization\""`
- Add a new postaction that runs before merge that, if the tag is defined, will:
-- Search for any files with names containing "$SEARCH$"
-- When it finds one it, it will search the other files in the same directory that have the same file extension and contain the text included in the tag value.
-- When it finds a file that contains the text, it will use that file's name to rename the one containing "$SEARCH$"
-- Allow normal merging to take care of the rest.

Possible issue/side-effects:
- This only supports a single rename action per template. I don't think that's going to be a problem for a while though. If it does become an issue this could be extended to support multiple search renamings per template. Or just avoided by splitting over multiple templates.
- Won't handle multiple "$SEARCH$" files as this is not needed. I can't imagine a situation where it would be that wouldn't also need different rename actions. (The point above)
- I'm also just planning on doing a simple `string.Contains()` test as I think this will be sufficient. If at some point we need to do a `Regex.Match()` test this can be changed or that functionality added as/when necessary.

I'm not handling these for now though as I don't want to waste time on things we may never need when there are lots of other things that we definitely need to do.

Will also need
- tests to verify new functionality
- Additional tests adding to the TemplateVerifier.

Can anyone see any problems with the above? Or know of an alternative approach (existing or already in development) that could be used instead?

**Comment by mrlacey** [comment]
Circular referencing was preventing chained references (A depends on B which depends on C) but have now changed this to detect direct circular referencing (A depends on B and B depends on A)
Will need to follow up and detect indirect circular referencing (A depends on B which depends on C which depends on A) up to any level of steps.
Would also be good to add testing for this in the TemplateValidator.

**Comment by ralarcon** [comment]
Hey @mrlacey,

We have been talking and with the rename postaction we see some risk with the right-click addition. What happen if I already have added a SettingsPage to my project during the "File -> New Project" and then I add a "Feedback hub" feature. Then the right-click generation must be aware of the exsiting page, so we need to articulate something new for this. A simple approach would be to do not allow the "Feedback hub" feature in the right-click.

Regarding circular references looks like a more complex algorithm is needed here.

**Comment by mrlacey** [comment]
Yes, I've had to disable right-click support for this feature.
It worked fine if no settings page was added. 
If added with the default name, it highlighted the conflicting files correctly. But,
If there was already a settings page with a different name, it added another one and so that's not acceptable.

Maybe if right-click support ever got more sophisticated there may be ways around this but enhancing the right-click merge/file-match logic is more than I can do for this feature.

**Comment by ralarcon** [comment]
It is possible to create a right-click post action to execute the "name" search and if so, use it. If not, means that the default should be added. By now we can have it disabled for right-click.

**Comment by mrlacey** [comment]
The challenge is not in the new postaction, the challenge is in the add wizard recognizing that an equivalent page with a different name already exists.

This issue already exists with adding pages. I can create an app with a settings page and then use right-click functionality to add another settings page as long as they have different names.
Single/Multi-Instance isn't handled by right-click functionality if names can change. Or, if names are fixed then it just creates a merge with no changes. Neither option is ideal.

I'll add the functionality without right-click support and maybe look to adjust the wizard later. (Possibly as part of this issue or I'll create a related one for that purpose.-It all depends on time.)
The more time I spend on the wizard means delays to other work I'd planned (e.g. VB support.)

**Comment by mrlacey** [comment]
This is done but waiting to see if I can get the changes necessary for right-click support in the same PR.

**Comment by mrlacey** [comment]
Submitted the PR without right-click support to make sure at least the add functionality makes it into this release. May look to add to this release if time allows. if not will raise a new issue to add the ability to detect the settings page, even if renamed, so it can be added at a later time. _Exact requirements to be spec'd_

**Comment by mrlacey** [comment]
Moving back to the next milestone as need longer to explore the best way to implement this.

**Comment by mrlacey** [comment]
can hopefully close out soon.

**Comment by mrlacey** [comment]
Verified in the dev-nightly build
